src/Couchbase.Lite.Shared/Query/QueryBase.cs

```diff
@@ -243,12 +185,35 @@ public ListenerToken AddChangeListener([NotNull] EventHandler<QueryChangedEventA
         public void RemoveChangeListener(ListenerToken token)
         {
             _disposalWatchdog.CheckDisposed();
-            _changed.Remove(token);
+            _listenerTokens[token].StopObserver();
+            _listenerTokens.Remove(token);
             if (Interlocked.Decrement(ref _observingCount) == 0) {
                 Stop();
             }
+        }
+
+        #endregion
+
+        #region QueryBase
 
-            _willUpdate.Set(false);
+        protected abstract void CreateQuery();
+
+        internal abstract unsafe Dictionary<string, int> CreateColumnNames(C4Query* query);
```

Comment:
Can you point out the need to make this an internal method?  What is the benefit of moving it here?  If none, it's better to keep it encapsulated like where it was before.  I don't see QueryBase calling it, or any other class calling it on QueryBase.

Reply:
It is used to get QueryResultSet in LiveQuerier class. Called in QueryObserverCalled(C4QueryObserver* obs, C4Query* query) method.

Reply:
I think this logic is still breaking outside of the class too much.  If you want to make sure ColumnNames is not calculated when you get it from inside of QueryBase, then how about just having a method called `ColumnNames()` that returns it directly, and calculates it if needed?